src/Microsoft.AspNetCore.Mvc.TagHelpers/Cache/DistributedCacheTagHelperService.cs

```diff
@@ -74,20 +77,62 @@ HtmlEncoder HtmlEncoder
 
                             value = await _formatter.SerializeAsync(formattingContext);
 
-                            await _storage.SetAsync(key, value, options);
+                            
+                            using (var buffer = new MemoryStream())
+                            {
+                                // The stored content is 
+                                // - Length of the serialized cache key in bytes
+                                // - Cache Key
+                                // - Content
+
+                                var keyLength = BitConverter.GetBytes(serializedKey.Length);
+
+                                await buffer.WriteAsync(keyLength, 0, keyLength.Length);
+                                await buffer.WriteAsync(serializedKey, 0, serializedKey.Length);
+                                await buffer.WriteAsync(value, 0, value.Length);
+
+                                await _storage.SetAsync(storageKey, buffer.ToArray(), options);
+                            }
 
                             content = formattingContext.Html;
                         }
                         else
                         {
-                            content = await _formatter.DeserializeAsync(value);
+                            // Extract the length of the serialized key
+                            byte[] contentBuffer = null;
+                            using (var buffer = new MemoryStream(value))
+                            {
+                                var keyLengthBuffer = new byte[sizeof(int)];
```

Comment:
Hmm, this would be a problem if in a farm you had a mix of x86/x64 processes running wouldn't it?